server.js

```diff
@@ -22,11 +14,98 @@ const app = express()
 app.use(cors())
 app.use(bodyParser.json())
 
-// Start defining your routes here
+
+// Root
 app.get('/', (req, res) => {
-  res.send('Hello world')
+  res.send('<pre>Paths:\n/books\n/books/:id\n/isbn/:isbn\n\nQueries:\n?author\n?title\n?language\n?minpages\n?maxpages\n?rating\n?sort\n?page\n?results<pre>')
+})
+
+
+// Books
+app.get('/books', (req, res) => {
+  const { title, author, language, minpages = 0, maxpages = Infinity, rating, sort, page, results = 20 } = req.query
+
+  const filterBooks = (array, filter) => {
+    return array.toString().toLowerCase().includes(filter)
+  }
+
+  let filteredBooks = booksData;
+
+  //Filter
+  if (author) {
+    filteredBooks = filteredBooks.filter(book => filterBooks(book.authors, author))
+  }
+  if (title) {
+    filteredBooks = filteredBooks.filter(book => filterBooks(book.title, title))
+  }
+  if (language) {
+    filteredBooks = filteredBooks.filter(book => filterBooks(book.language_code, language))
+  }
+  if (maxpages || minpages) {
+    filteredBooks = filteredBooks.filter(book => book.num_pages < +maxpages && book.num_pages > +minpages)
+  }
+  if (rating) {
+    filteredBooks = filteredBooks.filter(book => Math.round(book.average_rating) === +rating)
+  }
+
+  //Sort
+  //sorting on author and title was a lot more complicated that I assumed 
+  if (sort) {
+    if (sort === 'author') {
+      filteredBooks = filteredBooks.sort((a, b) => {
+        const authorA = a.authors.toLowerCase()
+        const authorB = b.authors.toLowerCase()
+        if (authorA < authorB) return -1;
+        if (authorA > authorB) return 1;
+        return 0;
+      });
+    } else if (sort === 'title') {
+      filteredBooks = filteredBooks.sort((a, b) => {
+        const titleA = a.title.toString().toLowerCase()
+        const titleB = b.title.toString().toLowerCase()
+        if (titleA < titleB) return -1;
+        if (titleA > titleB) return 1;
+        return 0;
+      });
```

Comment:
Two times the same logic, so maybe it would nicer to have it in a separate function and pass author and title as props?

Reply:
Good point, will do! Thanks!